These tutorials are intended to provide guidance for more sophisticated applications of \poy that involve multiple steps and a combination of different commands. Each tutorial contains a \poy script that is followed by detailed commentaries explaining the rationale behind each step of the analysis. Although these analyses can be conducted interactively using the \emph{Interactive Console} or running separate sequential analyses using the \emph{Graphical User Interphace}, the most practical way to do this is to use \poy scripts (see \emph{ POY4 Quick Start} for more information on \poy scripts).

It is important to remember that the numerical values for most command arguments will differ substantially depending on type, complexity, and size of the data. Therefore, the values used here should not be taken as optimal parameters.

The tutorials use sample datasets that are provided with \poy installation but can also be downloaded from the \poy site at
\begin{center}
\texttt{http://research.amnh.org/scicomp/projects/poy.php}
\end{center}
The minimal required items to run the tutorial analyses are the \poy application and the sample datafiles. Running these analyses requires some familiarity with \poy interface and command structure that can be found in the preceding chapters.

\section{Tutorial name}
Description of the tutorial.

\begin{verbatim}
(* my first tutorial *)
read()
build()
swap()
\end{verbatim}

\begin{description}
\item[read()] This command does...
\item[build()] The argument of this command specify...
\item[swap()] This command is necessary... 
\end{description}

\section{Basic Search}{\label{tutorial2}}

This tutorial illustrates the basics of running \poy using single input datafile. 
\begin{enumerate}
\item Navigate to the \texttt{tutorial} folder. 
\item Launch \poy (review \poy Quick Guide for instructions for your platform).
\item Type:

    \begin{verbatim}
    read ("mol1.txt") [enter]
    build (100) [enter]
    select (unique) [enter]
    swap (threshold:10) [enter]
    select () [enter]
    report (asciitrees) [enter]
    report ("tutorial1_trees.txt", trees) [enter]
    report ("tutorial1_stats.txt", treestats) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
\item  Import the DNA sequence datafile mol1.txt.
\item  Generate 100 random addition sequence Wagner trees.
\item  Discard duplicate trees.
\item  Alternate SPR and TBR branch swapping of all current trees as well as all 
new trees found that are up to 10\%
longer than the current tree being 
swapped. Note: Threshold during swapping is equivalent to slop in \texttt{POY3}. 
Given the faster algorithms and better heuristics of \poy it is likely to 
be less important than it was in \texttt{POY3}.
\item  Discard suboptimal and duplicate trees (i.e., retain only optimal trees).
\item  Draw optimal trees in \poy Output window (ncurses version only) or output 
file (non-ncurses version), reporting the cost of each tree.
\item  Output all current trees to file tutorial1\_trees.txt.
\item  Output basic tree statistics to file tutorial1\_stats.txt.
\end{itemize}

\item View cladogram(s) in parenthetical format by opening tutorial1\_trees.txt in 
your chosen text editor.
\item View basic tree statistics by opening tutorial1\_stats.txt in your chosen 
text editor.
\end{enumerate}

\section[Advanced Search]{Advanced Search I: Multiple Datasets and Data Types, Equal 
Weighting, Rooting, and Publication Quality Trees}{\label{tutorial2}}

This tutorial builds on Tutorial~\ref{tutorial1} to include multiple input datafiles and 
multiple data types (i.e., genotypic and phenotypic) using equal weighting, 
designate the root, and automatically generate publication quality trees.

\begin{enumerate}
\item Type: 

\begin{verbatim}
    read ("mol2.txt", "morph.txt") [enter]
    set (root: "Hagfish") [enter]
    transform ((all, tcm:(1,1))) [enter]
    build (100) [enter]
    select (unique) [enter]
    swap (threshold:10) [enter]
    select () [enter]
    report ("tutorial2", graphtrees) [enter]
    report ("tutorial2_trees.txt", trees) [enter]
    report ("tutorial2_stats.txt", treestats) [enter]
\end{verbatim}

The above commands will perform the following tasks using default parameters:

\begin{itemize}
\item Import the DNA sequence datafile mol2.txt and the phenotypic datafile 
morph.ss. 
\item Designate the Hagfish as the root. 
\item Set indel and substitution costs to 1 (equal weighting). The additivity 
(ordering) of the phenotypic characters is determined in the ccode of the 
NONA file morph.ss. By default, the transformation cost for phenotypic 
characters is 1.  
\item Generate 100 random addition sequence Wagner trees.
\item Discard duplicate trees.
\item For each tree in memory, alternate SPR and TBR branch swapping of the tree 
and all trees found within 10\% of the cost of the tree.
\item Discard suboptimal and duplicate trees (\emph{i.e.}, retain only optimal trees).
\item Output publication-quality cladogram(s) of optimal tree(s) to a postscript 
file tutorial2.ps. (Note: \poy automatically adds the \texttt{.ps} extension.)  
\item Output all current trees to file tutorial2\_trees.txt.
\item Output basic tree statistics to file tutorial2\_stats.txt.
\end{itemize}

\item View publication-quality cladogram(s) by opening tutorial2.ps in your chosen 
postscript viewer.
\item View cladogram(s) in parenthetical format by opening tutorial2\_trees.txt in 
your chosen text editor.
\item View basic tree statistics by opening tutorial2\_stats.txt in your chosen 
text editor.

\end{enumerate}

\section[Fusing and Ratcheting]{Advanced Search II: Tree Fusing and Fragment
Ratcheting}{\label{tutorial3}}

This tutorial builds on Tutorial~\ref{tutorial2} to use tree fusing and fragment ratcheting 
that help to escape suboptimal islands. This tutorial also introduces matrix and 
tree output in NONA format.

\begin{enumerate}
\item Type: 
\begin{verbatim}
    build (100) [enter]
    select (unique) [enter]
    fuse () [enter]
    select () [enter]
    perturb (iterations:10, ratchet:(0.2,5)) [enter]
    select () [enter]
    swap (threshold:10) [enter]
    select () [enter]
    report ("tutorial3", graphtrees) [enter]
    report ("alignment3.ss", phastwinclad) [enter]
\end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
\item Generate 100 random addition sequence Wagner trees.
\item Discard duplicate trees.
\item Perform cladogram searching by fusing one pair of trees in memory. Note: Tree 
fusing requires at least two trees; if there was only one tree in memory, 
an error message will be generated. 
\item Discard suboptimal and duplicate trees (\emph{i.e.}, retain only optimal trees).
\item Perform 10 successive repetitions of a fragment-based ratchet by randomly 
selecting 20\% of the fragments and upweighting them by a factor of five.
\item Discard suboptimal and duplicate trees (i.e., retain only optimal trees).
\item For each tree in memory, alternate SPR and TBR branch swapping of the tree 
and all trees found within 10\% of the cost of the tree.
\item Discard suboptimal and duplicate trees (\emph{i.e.}, retain only optimal trees).
\item Output publication-quality cladogram(s) of optimal tree(s) to a postscript 
file tutorial3.ps. (Note: \poy automatically adds the \texttt{.ps} extension.)
\item Output NONA file with the implied alignment for one (if multiple) optimal 
trees and all optimal trees in the file alignment3.ss.
\end{itemize}

\item View publication-quality cladogram(s) by opening tutorial3.ps in your chosen 
postscript viewer.
\item View implied alignment, optimal tree(s), and manipulate data by opening the 
file alignment3.ss in WinClada or MacClade.

\end{enumerate}

\section[Trees and Step Matrices]{Advanced Search III: Input Trees and Step
Matrices}{\label{tutorial4}}

This tutorial builds on Tutorials~\ref{tutorial2} and~\ref{tutorial3}, and introduces the use of input trees 
and complex differential cost step matrices.

\begin{enumerate}
\item Type: 
    \begin{verbatim}
    read ("trees.txt") [enter]
    transform ((all, tcm:"g4ts1tv2.txt"), 
    (static, weightfactor:4)) [enter]
    swap (threshold:10) [enter]
    select () [enter]
    report ("tutorial5", graphtrees) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
\item Read cladograms from the input file trees.txt.
\item Apply specified transformation costs to data. The transformation cost 
matrix in the file g4ts1tv2.txt is applied to the dynamic homology 
characters (i.e., unaligned DNA sequences) to assign indel events a cost 
of 4, transitions 1, and transversions 2. All static homology characters 
(i.e., the phenotypic characters in the file morph.txt) are upweighted by 
a factor of 4, with additivities and previous relative weights unchanged 
(e.g., a character with transformations costs of 2 would be now have a 
transformation cost of 8).
\item For each tree in memory, alternate SPR and TBR branch swapping of the tree 
and all trees found within 10\% of the cost of the tree.
\item Discard suboptimal and duplicate trees (i.e., retain all most optimal 
trees).
\item Output publication-quality cladogram(s) of optimal tree(s) to a postscript 
file tutorial5.ps. (Note: \poy automatically adds the .ps extension.)  
\end{itemize}

\item View publication-quality cladogram(s) by opening tutorial4.ps in your chosen 
postscript viewer.
\item View cladogram(s) in parenthetical format by opening tutorial4\_trees.txt in 
your chosen text editor.
\item View basic tree statistics by opening tutorial4\_stats.txt in your chosen 
text editor.
\end{enumerate}

\section[Bremer Support]{Support I: Bremer Support}

This tutorial builds on the previous tutorials to illustrate Bremer support 
calculation.
\begin{enumerate}
\item Type: 
    \begin{verbatim}
    transform ((all, tcm:(1,1)), (static, weightfactor:1)) [enter]
    swap () [enter]
    calculate_support (bremer, build (trees:5), swap (trees:2)) [enter]
    report (supports) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\item Apply a weight of one to all transformations.
\item For each tree in memory, alternate SPR and TBR branch swapping of the 
optimal tree(s).
\item Estimate Bremer support by using inverse constraints, doing five 
independent searches for every group, holding a maximum of two trees. 
\item Output support values for each group in parenthetical notation to \poy 
output window.

\end{enumerate}

\section[Bootstrap Support]{Support II: Bootstrap Support Using Dynamic Homology}

This tutorial builds on the previous tutorials to illustrate the calculation of 
bootstrap frequencies. As discussed in the \poy Commands Reference, the 
characters sampled during pseudoreplicates are entire fragments of DNA 
sequences, not individual nucleotide characters. Tutorial 7 shows how to 
estimate bootstrap frequencies using static homology, which allows nucleotide-
level characters to be sampled. 

\begin{enumerate}
\item Type:
    \begin{verbatim}
    calculate_support (bootstrap: 100, build(trees:2), 
    swap(trees:1)) [enter]
    report (supports) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
\item Perform 100 pseudoreplicates by sampling characters with replacement, 
doing two independent searches for each pseudoreplicate and holding a 
maximum of one tree. 
\item Output bootstrap frequencies for each group in parenthetical notation to 
\poy output window.
\end{itemize}
\end{enumerate}

\section[Bootstrap Support with Static Homologies]{Support III: Bootstrap Support Using Static Homology}

This tutorial builds on the previous tutorials to illustrate the calculation of 
bootstrap frequencies. Here, bootstrap frequencies are obtained from analysis of 
the implied alignment of static homologies, which permits individual nucleotide-
level characters to be sampled instead of whole fragments, as is done using 
dynamic homology. 
\begin{enumerate}
\item Type:
    \begin{verbatim}
    transform ((all, static_approx)) [enter]
    calculate_support (bootstrap: 100, build(trees:2), 
    swap(trees:1)) [enter]
    report (supports) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
\item Generate the alignment implied by the optimal tree and given the assumed 
transformation costs. 
\item Perform 100 pseudoreplicates by sampling characters with replacement, 
doing two independent searches for each pseudoreplicate and holding a 
maximum of one tree. 
\item Output bootstrap frequencies for each group in parenthetical notation to 
\poy output window.
\end{itemize}
\end{enumerate}

\section[Chromosome Analysis]{Chromosome Analysis I: Unannotated Sequences}

This tutorial illustrates the analysis of chromosome-level transformations using 
unannotated sequences, i.e., contiguous strings of sequences without prior 
identification of independent regions.
\begin{enumerate}
    \item Type
    \begin{verbatim}
    wipe () [enter]
    read (chromosome:("mit5.txt"))  [enter]
    transform ((all, dynamic_pam:(inversion:15, locus_indel:(10, 1.5), 
    median:  3, swap_med:5, circular:true, approx:true))) [enter]
    build (5) [enter]
    swap ()[enter]
    select () [enter]
    report (asciitrees, diagnosis) [enter]
    transform ((all, dynamic_pam:(inversion:15, locus_indel:(10, 1.5), 
    median:3, swap_med:5, circular:true, approx:false))) [enter]
    swap ()[enter]
    select ()[enter]
    report (asciitrees, diagnosis) [enter]
    \end{verbatim}

The above commands will perform the following tasks using default parameters:
\begin{itemize}
    \item Clear all data and trees from memory.
    \item Import datafile mit5.txt.
    \item Treat all data as pertaining to unannotated chromosome data, setting the 
        following parameters: inversion distance and cost 15, locus indel 10 + 1.5 
        times the length (number of nucleotides) of the locus, keep three 
        candidate medians, swap on medians for 5 rounds, treat as circular 
        chromosome, and use fixed-states optimization to approximate chromosome 
        medians.
    \item Generate 5 random addition sequence Wagner trees.
    \item Alternate SPR and TBR branch swapping of each tree in memory.
    \item Discard suboptimal and duplicate trees (i.e., retain only optimal trees).
    \item Draw optimal trees in \poy Output window (ncurses version only) or output 
        file (non-ncurses version), reporting the cost of each tree.
    \item Output the optimal median states and edge costs.
    \item Treat all data as pertaining to unannotated chromosome data with 
        parameters as above but using optimization alignment (not fixed-states) to 
        approximate chromosome medians.   
    \item Alternate SPR and TBR branch swapping of each tree in memory.
    \item Draw optimal trees in \poy Output window (ncurses version only) or output 
        file (non-ncurses version), reporting the cost of each tree.
    \item Output optimal median states and edge costs.
\end{itemize}
\end{enumerate}
